### ISSUE TITLE
ProjectTo InvalidOperationException

### ISSUE THREAD

**Issue by thomaslevesque** [AutoMapper/AutoMapper#3432]
### Source/destination types

```csharp
// EF Core entities

public class Foo
{
    public int Id { get; set; }
    public ICollection<FooTranslation> Translations { get; set; } = new List<FooTranslation>();
    // other properties, not relevant to this issue
}

public class FooTranslation
{
    public int FooId { get; set; }
    public string Language { get; set; }
    public string Name { get; set; }
    public string Description { get; set; }
}

// DTOs
public class FooDto
{
    public int Id { get; set; }
    public LocalizedText Name { get; set; }
    public LocalizedText Description { get; set; }
    // other properties, not relevant to this issue
}

// The dictionary key is the language
public class LocalizedText : Dictionary<string, string> { }
```

### Mapping configuration

```csharp
CreateMap<Foo, FooDto>()
    .ForMember(dto => dto.Name, o => o.MapFrom(f => f.Translations.ToLocalizedText(t => t.Name)))
    .ForMember(dto => dto.Description, o => o.MapFrom(f => f.Translations.ToLocalizedText(t => t.Description)))
```

`ToLocalizedText` is a method that converts a list of `FooTranslation` to a `LocalizedText` (I can post the code if requested, but it's probably not relevant)

### Version
AutoMapper 9.0.0
EF Core 3.1.3

### Expected behavior

`dbContext.Foos.ProjectTo<FooDto>(configurationProvider)` should work

### Actual behavior

It throws an exception:

> System.InvalidOperationException: When called from 'VisitLambda', rewriting a node of type 'System.Linq.Expressions.ParameterExpression' must return a non-null value of the same type. Alternatively, override 'VisitLambda' and change it to not visit children of this type.

### Steps to reproduce

Project Foo to FooDto in a EF query:

```csharp
var dtos = await _dbContext.Foos.Project<FooDto>(_mapper.ConfigurationProvider);
```

---

I realize that it looks like an EF issue rather than an AutoMapper issue, but bear with me for a moment.

The problem is probably that EF Core can't translate `ToLocalizedText` to SQL, which makes sense. But I *have* to use `ToLocalizedText`, because if I just do `MapFrom(f => f.Translations)`, it doesn't return the right thing.

Basically, what I need is a way to specify that `Name` is mapped from `Translations`, so that the query has the right `Include`s, and also specify that the value should be converted using `ToLocalizedText`. But I can't put `ToLocalizedText` in the expression, since it can't translate to SQL.

Is there a way to do that? i.e., specify separately:
- where the value comes from (`Translations`)
- how to convert it to the destination type

I also tried this:

```csharp
.ForMember(dto => dto.Name, o => o.ConvertUsing(
    new TranslationValueConverter<FooTranslation>(t => t.Name),
    f => f.Translations));
```
But it returns a different error: 

> System.Exception: Can't resolve this to Queryable Expression

Am I just missing something obvious?

**Comment by thomaslevesque** [comment]
Basically, what I'm looking for is a `MapFrom` overload which lets me specify a conversion separately from the source member, i.e. something like this:

```
.ForMember(dto => dto.Name, o => o.MapFrom(
    f => f.Translations,
    translations => translations.ToLocalizedText(t => t.Name)));
```
The first argument would be used for the Linq projection, the second would be applied after the results have been fetched from the DB.

Is there already something in AutoMapper that would let me achieve this? Or is it a feature you'd consider adding?

**Comment by jbogard** [comment]
This is not a feature that's possible to add to AutoMapper. Once we give the projection to the query provider, execution of the LINQ query is completely in the LINQ provider's hands:

```csharp
   var foos = await _dbContext
        .Where(foo => foo.Bar > 3) // System.Linq.QueryableExtensions
        .ProjectTo<FooDto>() // returns IQueryable
        .ToListAsync(); // executed by the IQueryable.Provider
```

Once you go `IQueryable`, your choices for intercepting or extending the query execution are completely determined by the `IQueryable.Provider`. Once execution is *finished*, then you can do post-processing on the server-side if you like (after `ToListAsync`).

If AutoMapper were to support what you suggest, we'd have to wrap query providers, execute their result, then transform the result and assign values once it's finished. And because each query provider is different, for N query providers we'd have to have N specific packages to that query provider.

We can do `ProjectTo` because ALL that's doing is calling `Select` on the underlying query provider. 

In your code, I'd suggest doing your own post-processing of the value after calling `ToList`.

**Comment by thomaslevesque** [comment]
Thanks for the reply, @jbogard. I see the problem now.

> In your code, I'd suggest doing your own post-processing of the value after calling `ToList`.

Yes, that's what I ended up doing.
